### ISSUE TITLE
Invalid state when using %unit% in items and restoreOnStartup persistence strategy

### ISSUE THREAD

**Issue by wborn** [eclipse/smarthome#5675]
It seems that when UoM was introduced the unit is not saved for persisted items in legacy persistence addons in openhab1-addons. So when using the %unit% in items this results in an invalid state of items when they are restored on startup.

It can be reproduced with mapdb and the following mapdb.persist:

```php
Strategies {
    default = everyUpdate
}

Items {
    * : strategy = everyChange, restoreOnStartup
}
```

Then use the following test.items:

```php
Number:Energy Energy "Energy [%.1f %unit%]" <energy>
Number:Power Power "Power [%.1f %unit%]" <poweroutlet_eu>
Number:Dimensionless Humidity "Humidity [%.1f %unit%]" <humidity>
Number:Temperature Temperature "Temperature [%.1f %unit%]" <temperature>
```

Update the state on the Console and check the item state:

```
openhab> smarthome:update Energy "1.0 kWh" && smarthome:update Power "100.0 W" && smarthome:update Humidity "50 \%" && smarthome:update Temperature "20.0 °C"                                                                          
Update has been sent successfully.
Update has been sent successfully.
Update has been sent successfully.
Update has been sent successfully.

openhab> items list
Energy (Type=NumberItem, State=1.0 kWh, Label=Energy, Category=energy)
Power (Type=NumberItem, State=100.0 W, Label=Power, Category=poweroutlet_eu)
Humidity (Type=NumberItem, State=50 %, Label=Humidity, Category=humidity)
Temperature (Type=NumberItem, State=20.0 ℃, Label=Temperature, Category=temperature)
```

After restarting the runtime the item state is as follows:
```
openhab> items list
Energy (Type=NumberItem, State=1.0, Label=Energy, Category=energy)
Power (Type=NumberItem, State=100.0, Label=Power, Category=poweroutlet_eu)
Humidity (Type=NumberItem, State=50.0, Label=Humidity, Category=humidity)
Temperature (Type=NumberItem, State=20.0 ℃, Label=Temperature, Category=temperature)
```

In Basic UI this shows as:

![state](https://user-images.githubusercontent.com/12213581/40892272-17d70c78-6795-11e8-932f-c1ec745afdb1.png)

A workaround for this is to replace %unit% with real units:

```php
Number:Energy Energy "Energy [%.1f kWh]" <energy>
Number:Power Power "Power [%.1f W]" <poweroutlet_eu>
Number:Dimensionless Humidity "Humidity [%.1f %%]" <humidity>
Number:Temperature Temperature "Temperature [%.1f °C]" <temperature>
```

Most of  the documenation is now using %unit% so this will be a common issue for people using the restoreOnStartup persistence strategy.

**Note that there is an issue in the mapdb persistence openhab1-addon causing the state of items to be not immediately stored (https://github.com/openhab/openhab1-addons/pull/5591).**

I've tested this with openHAB 2.4.0-SNAPSHOT (# 1294).

**Comment by lolodomo** [comment]
Strange that it is ok with °C and not other units.

**Comment by wborn** [comment]
It only works for units in the [dimensionMap of the I18nProviderImpl](https://github.com/eclipse/smarthome/blob/84e188da8142cc885e881a551b54859df452fcb2/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/i18n/I18nProviderImpl.java#L345). I looks like these units are always used with %unit% items, also  when the unit is missing.

**Comment by lolodomo** [comment]
We probably need the analysis of @htreu, our UoM specialist

**Comment by htreu** [comment]
@wborn's analysis is correct. The units are not persisted and on restore the default units are taken from the `I18nProviderImpl` since the items do not specifically define a unit. As `Temperature` does define `°C/°F` as defaults and `Dimensionless` defines the abstract unit `ONE`  the warnings in the log only show up for `Energy` and `Power`. Nevertheless does `Humidity` lack the correct unit.
We have the MapDBPersistenceService in the pipeline with #5419 which should again be checked to be compatible with uom. It is meant to be a "restore-only" persistence which would address this issue.

**Comment by lolodomo** [comment]
You mean we will have to trash rrdj persistence and replace it with this new persistence service ?

**Comment by kaikreuzer** [comment]
Hopefully not, UoM was meant to be backward compatible...

**Comment by htreu** [comment]
No, the new MapDBPersistenceService serves a different purpose. See #5419 for discussion.

**Comment by kaikreuzer** [comment]
Why a different purpose? They all should just be implementations of the PersistenceService interface and rrd4j supports the "restoreOnStartup" strategy as well (as many others do).

**Comment by htreu** [comment]
> If the value was in Kelvin, after restart, it will be in °C.

That depends on the item definition. You should make sure to define the unit in the item's state description. This way the value restored from the persistence service will again have the unit defined on the item. A state description using the generic placeholder `%unit%` suffers from the issue described here.

> That looks like a major issue that shows that UoM is incompatible with any OH1 persistence service.

OH1 persistence services do not support the full generic unit definition (channel provided unit vs. system default unit). Persistence services should be migrated which was out-of-scope when UoM was introduced.

**Comment by lolodomo** [comment]
Do you mean that if I define my item like that
Number:Temperature temp label="Température [%.1f K]" { channel="xxx" }
and if the binding is setting a value using °C,
then after restart the value will be correct in K ?
If that's true, we have to advice users to not use %unit% in item definition.

**Comment by kaikreuzer** [comment]
> we have to advice users to not use %unit% in item definition.

Correct, that is the workaround for persistence services that do not support UoM: Simply fix the desired unit in the item definition.

**Comment by wborn** [comment]
Perhaps the PersistanceManagerImpl can ignore and warn in [initialize(Item item)](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/internal/PersistenceManagerImpl.java#L249) when HistoricItems are restored that should have a dimension but which is missing?
